### PR TITLE
[Fixes #167] Error page/handler should reset any buffered data, clear …

### LINKED ISSUE #167: Error page/handler should reset any buffered data, clear headers, etc.
RE: https://github.com/aspnet/Hosting/issues/299

If the response is buffering when an exception occurs, the error handler and error page middleware should clear the headers and buffer before taking their recovery actions.

ErrorHandler already clears the headers and sets a status code:
https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics/ErrorHandlerMiddleware.cs#L59-L61

ErrorPage sets the status code but only alters select headers. It doesn't even check Response.HasStarted before trying to do anything, so in this scenario it is already throwing trying to modify the status code and headers.
https://github.com/aspnet/Diagnostics/blob/dev/src/Microsoft.AspNet.Diagnostics/Views/ErrorPage.cshtml#L16-L19

### PR DESCRIPTION
**Pull request by kichalla**
@Tratcher (fyi...i installed the locally built package of buffering and tested it)

### CODE REVIEW

File: src/Microsoft.AspNet.Diagnostics/ErrorHandlerMiddleware.cs [hunk 1]
```diff
@@ -60,7 +61,16 @@ public ErrorHandlerMiddleware(RequestDelegate next, ILoggerFactory loggerFactory
                     context.Response.Headers.Clear();
                     context.Response.OnStarting(_clearCacheHeadersDelegate, context.Response);
 
-                    // TODO: Try clearing any buffered data. The buffering feature/middleware has not been designed yet.
+                    // if buffering is enabled, then clear it as data could have been written into it.
+                    var bufferingFeature = context.GetFeature<IHttpBufferingFeature>();
+                    if (bufferingFeature != null)
+                    {
+                        if (context.Response.Body.CanSeek)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Diagnostics/ErrorHandlerMiddleware.cs]: I wonder if we should even check for the presence of `IHttpBufferingFeature` and just do this check?

Comment (Tratcher) [inline src/Microsoft.AspNet.Diagnostics/ErrorHandlerMiddleware.cs]: No, you don't need to check for the IHttpBufferingFeature, just CanSeek

File: test/Microsoft.AspNet.Diagnostics.Tests/project.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
         "warningsAsErrors": "true"
     },
     "dependencies": {
+        "Microsoft.AspNet.Buffering": "1.0.0",
```
Comment (Tratcher) [inline test/Microsoft.AspNet.Diagnostics.Tests/project.json]: If you check this dependency in we'll have to add the dependency in the CI graph.

**Comment by Tratcher** [comment]
:watch:

**Comment by kichalla** [comment]
updated @Tratcher

File: src/Microsoft.AspNet.Diagnostics/Views/ErrorPage.cshtml [hunk 3]
```diff
@@ -13,10 +13,8 @@
     public ErrorPageModel Model { get; set; }
 }
 @{
-    Response.StatusCode = 500;
     // TODO: Response.ReasonPhrase = "Internal Server Error";
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Diagnostics/Views/ErrorPage.cshtml]: Remove this line.

**Comment by Tratcher** [comment]
:shipit:

**Comment by kichalla** [comment]
09b1c0fa6314db316b386c166e07b80cc27ed0c2

**Pull request closed without merging**